crates/lib/src/execution_data.rs

```diff
@@ -18,3 +23,192 @@ pub struct ExecutionData {
     /// This will always be `None` for QVM execution.
     pub duration: Option<Duration>,
 }
+
+#[derive(Debug, Clone, PartialEq)]
+#[repr(transparent)]
+pub struct ReadoutMap(HashMap<MemoryReference, ReadoutValues>);
+
+impl ReadoutMap {
+    /// Given a known readout field name and index, return the result's ReadoutValues, if any.
+    pub fn get_readout_values(&self, field: &str, index: u64) -> Option<ReadoutValues> {
+        let readout_values = self.0.get(&MemoryReference {
+            name: String::from(field),
+            index,
+        })?;
+
+        Some(readout_values.clone())
+    }
+
+    /// Given a known readout field name, return the result's ReadoutValues for all indices, if any.
+    pub fn get_readout_values_for_field(
+        &self,
+        field: &str,
+    ) -> Result<Option<Vec<Option<ReadoutValues>>>, TryFromIntError> {
+        let mut readout_values = Vec::new();
+        for (memref, values) in &self.0 {
+            let MemoryReference { name, index } = memref;
+            let index = usize::try_from(*index)?;
+            if name == field {
+                if readout_values.len() <= index {
+                    readout_values.resize(index + 1, None);
+                }
+                readout_values[index] = Some(values.clone());
+            }
+        }
+
+        Ok(match readout_values.len() {
+            0 => None,
+            _ => Some(readout_values),
+        })
```

Comment:
Optional, just a couple alternative ways to write this.

```suggestion
        Ok((!readout_values.is_empty()).then_some(readout_values))
```

```suggestion
        Ok(if readout_values.is_empty() {
            None
        } else {
            Some(readout_values)
        })
```